src/main/java/io/appform/dropwizard/actors/base/UnmanagedPublisher.java

```diff
@@ -106,6 +113,79 @@ public final long pendingSidelineMessagesCount() {
         return Long.MAX_VALUE;
     }
 
+    /**
+     * Note: Timeout is in MilliSeconds, Function take a list of message as input and return not ack msg as output
+     * @param messages : Messages to be published
+     * @param properties
+     * @param timeout : in MS timeout for waiting on countDownLatch
+     * @param unit : timeout unit
+     * @return : List of message nacked
+     * @throws Exception
+     */
+    public final List<Message> publishWithConfirmListener(List<Message> messages, AMQP.BasicProperties properties,
+            long timeout, @NotNull TimeUnit unit) throws Exception {
+        publishChannel.confirmSelect();
+        ConcurrentNavigableMap<Long, Message> outstandingConfirms = new ConcurrentSkipListMap<>();
+        final CountDownLatch publishAckLatch = new CountDownLatch(messages.size());
+        publishChannel.addConfirmListener((sequenceNumber, multiple) -> {
+            if (multiple) {
+                ConcurrentNavigableMap<Long, Message> confirmed = outstandingConfirms.headMap(
+                        sequenceNumber, true
+                );
+                for(int i =0;i<confirmed.size();i++)
+                    publishAckLatch.countDown();
+                confirmed.clear();
+            } else {
+                publishAckLatch.countDown();
+                outstandingConfirms.remove(sequenceNumber);
+            }
+            log.info(String.format("Message with delivery tag %s acknoledged", sequenceNumber));
+        }, (sequenceNumber, multiple) -> {
+            {
+                if(multiple == true)
+                {
+                    ConcurrentNavigableMap<Long, Message> confirmed = outstandingConfirms.headMap(
+                            sequenceNumber, true
+                    );
+                    for(int i =0;i<confirmed.size();i++)
+                        publishAckLatch.countDown();
+                }
+                else
+                    publishAckLatch.countDown();
+            }
+            log.debug(String.format("Message nacked with multiple= %s : %s", multiple, outstandingConfirms.get(sequenceNumber)));
+        });
+
+        String routingKey;
+        if (config.isSharded()) {
+            routingKey = NamingUtils.getShardedQueueName(queueName, getShardId());
+        } else {
+            routingKey = queueName;
+        }
```

Comment:
Move this routingKey calculation to a method

Reply:
done